### PR TITLE
Correctly handle emtpy string in proxyuserpwd config

### PR DESCRIPTION
**Pull request by sshambar**
As documented, the default value for config value proxyuserpwd is ''.
However, that value results in the error:
 "cURL error 5: Unsupported proxy syntax in '@'".
This patch handles the values of '' and null (the default in the code)
the same for config values proxyuserpwd and proxy.

Signed-off-by: Scott Shambarger <devel@shambarger.net>

### CODE REVIEW

File: lib/private/Http/Client/Client.php [hunk 1]
```diff
@@ -97,13 +97,13 @@ private function getCertBundle(): string {
 	private function getProxyUri(): ?string {
 		$proxyHost = $this->config->getSystemValue('proxy', null);
 
-		if ($proxyHost === null) {
+		if (!$proxyHost) {
```
Comment (rullzer) [inline lib/private/Http/Client/Client.php]:
```suggestion
		if ($proxyHost !== '') {
```

File: lib/private/Http/Client/Client.php [hunk 2]
```diff
@@ -97,13 +97,13 @@ private function getCertBundle(): string {
...
 			return null;
 		}
 
 		$proxyUserPwd = $this->config->getSystemValue('proxyuserpwd', null);
 
-		if ($proxyUserPwd === null) {
+		if (!$proxyUserPwd) {
```
Comment (rullzer) [inline lib/private/Http/Client/Client.php]:
```suggestion
		if ($proxyUserPwd !== '') {
```

File: lib/private/Http/Client/Client.php [hunk 3]
```diff
@@ -97,13 +97,13 @@ private function getCertBundle(): string {
 	private function getProxyUri(): ?string {
 		$proxyHost = $this->config->getSystemValue('proxy', null);
```
Comment (rullzer) [inline lib/private/Http/Client/Client.php]:
```suggestion
		$proxyHost = $this->config->getSystemValue('proxy', '');
```

**Review by rullzer — COMMENTED** [review]
This is cleaner IMO. since it sets the correct default

File: lib/private/Http/Client/Client.php (hunk 3 above)
Comment (sshambar) [reply lib/private/Http/Client/Client.php]: While technically more correct, I was concerned about anyone whose used the existing "code default" of null, and will break with the suggested change (null in config.php would result in an error, where it would work currently).

I thought it better to use the example of Server.php:507 calling Memcache/Factory.php:74 as it works for both '' and null values.

But I'm willing to defer to the more strict approach if it gets a patch accepted :)

Thanks,
Scott

Comment (rullzer) [reply lib/private/Http/Client/Client.php]: Mmmm fair enough... I just don't like `!$foo` since in PHP many unexpected things can be false ;)
Maybe add an extra check. So change it to correct default. But also do an explicit test for null ;)?

Comment (sshambar) [reply lib/private/Http/Client/Client.php]: I've updated the patch to explicitly check for null or ''

**Review by rullzer — APPROVED** [review]
Awesome :)

**Comment by kesselb** [comment]
https://github.com/nextcloud/server/blob/effca30d1057deb56ed181662ffc3549ddca06e5/tests/lib/Http/Client/ClientTest.php#L56-L65

@rullzer do you know why these tests are not failing? `getSystemValue` is invoked with a different default value :thinking:

**Comment by rullzer** [comment]
> https://github.com/nextcloud/server/blob/effca30d1057deb56ed181662ffc3549ddca06e5/tests/lib/Http/Client/ClientTest.php#L56-L65
> 
> @rullzer do you know why these tests are not failing? `getSystemValue` is invoked with a different default value thinking

I think because null is not matched exactly in the with statement...

**Review by ChristophWurst — APPROVED** [review]
`empty` might also have done the trick but this is fine :)

Thanks a lot!

**Comment by ChristophWurst** [comment]
> I think because null is not matched exactly in the with statement...

one could then try

```php
$this->config 
 			->expects($this->at(1)) 
 			->method('getSystemValue') 
 			->with($this->equalTo('proxyuserpwd'), $this->equalTo(null)) 
 			->willReturn(null); 
```

That might help in case phpunit filters out empty args.

**Comment by rullzer** [comment]
Ah my bad. pushed wrong. let me create a proper branch.
